From dwmw2@infradead.org  Tue May  5 09:25:45 2009
From: David Woodhouse <dwmw2@infradead.org>
Date: Tue, 05 May 2009 09:25:26 +0100
Subject: intel-iommu: Fix oops in device_to_iommu() when devices not found.
To: stable@kernel.org
Message-ID: <1241511926.6126.188.camel@macbook.infradead.org>

From: David Woodhouse <dwmw2@infradead.org>

(cherry picked from commit 4958c5dc7bcb2e42d985cd26aeafd8a7eca9ab1e)

It's possible for a device in the drhd->devices[] array to be NULL if
it wasn't found at boot time, which means we have to check for that
case.

Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/pci/intel-iommu.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/pci/intel-iommu.c
+++ b/drivers/pci/intel-iommu.c
@@ -452,7 +452,8 @@ static struct intel_iommu *device_to_iom
 			    drhd->devices[i]->bus->number == bus &&
 			    drhd->devices[i]->devfn == devfn)
 				return drhd->iommu;
-			if (drhd->devices[i]->subordinate &&
+			if (drhd->devices[i] &&
+			    drhd->devices[i]->subordinate &&
 			    drhd->devices[i]->subordinate->number <= bus &&
 			    drhd->devices[i]->subordinate->subordinate >= bus)
 				return drhd->iommu;
